src/script/miniscript.h

```diff
@@ -0,0 +1,998 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SCRIPT_MINISCRIPT_H
+#define BITCOIN_SCRIPT_MINISCRIPT_H
+
+#include <algorithm>
+#include <numeric>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include <stdlib.h>
+#include <assert.h>
+
+#include <script/script.h>
+#include <span.h>
+#include <util/spanparsing.h>
+#include <util/strencodings.h>
+#include <util/vector.h>
+#include <primitives/transaction.h>
+
+namespace miniscript {
+
+/** This type encapsulates the miniscript type system properties.
+ *
+ * Every miniscript expression is one of 4 basic types, and additionally has
+ * a number of boolean type properties.
+ *
+ * The basic types are:
+ * - "B" Base:
+ *   - Takes its inputs from the top of the stack.
+ *   - When satisfied, pushes a nonzero value of up to 4 bytes onto the stack.
+ *   - When dissatisfied, pushes a 0 onto the stack.
+ *   - This is used for most expressions, and required for the top level one.
+ *   - For example: older(n) = <n> OP_CHECKSEQUENCEVERIFY.
+ * - "V" Verify:
+ *   - Takes its inputs from the top of the stack.
+ *   - When satisfactied, pushes nothing.
+ *   - Cannot be dissatisfied.
+ *   - This is obtained by adding an OP_VERIFY to a B, modifying the last opcode
+ *     of a B to its -VERIFY version (only for OP_CHECKSIG, OP_CHECKSIGVERIFY
+ *     and OP_EQUAL), or using IFs where both branches are also Vs.
```

Comment:
I've updated the comment to be more general but not exhaustive:
```diff
diff --git a/src/script/miniscript.h b/src/script/miniscript.h
index 88270e54d7..070dcc9d14 100644
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -42,9 +42,9 @@ namespace miniscript {
  *   - Takes its inputs from the top of the stack.
  *   - When satisfactied, pushes nothing.
  *   - Cannot be dissatisfied.
- *   - This is obtained by adding an OP_VERIFY to a B, modifying the last opcode
+ *   - This can be obtained by adding an OP_VERIFY to a B, modifying the last opcode
  *     of a B to its -VERIFY version (only for OP_CHECKSIG, OP_CHECKSIGVERIFY
- *     and OP_EQUAL), or using IFs where both branches are also Vs.
+ *     and OP_EQUAL), or by combining a V fragment under some conditions.
  *   - For example vc:pk_k(key) = <key> OP_CHECKSIGVERIFY
  * - "K" Key:
  *   - Takes its inputs from the top of the stack.
```